### PR TITLE
Fix redundant dispatch

### PR DESCRIPTION
**Pull request by ivanmoskalev**
Working on issue #78 

- Made `Bond` Equatable on the basis of memory address
- `BondBox` infers equality from the `Bond` it wraps
- Ref-ing `Bonds` in `Dynamic` with a `Set` (fixes redundant dispatch)
- Added performance tests for relevant functions

Before:
```
Test Case '-[BondTests.BondPerformanceTests testBindPerformance]' measured [Time, seconds] average: 0.000, relative standard deviation: 124.554%, values: [0.000041, 0.000010, 0.000011, 0.000003, 0.000006, 0.000003, 0.000003, 0.000002, 0.000006, 0.000004], performanceMetricID:com.apple.XCTPerformanceMetric_WallClockTime, baselineName: "", baselineAverage: , maxPercentRegression: 10.000%, maxPercentRelativeStandardDeviation: 10.000%, maxRegression: 0.100, maxStandardDeviation: 0.100
Test Case '-[BondTests.BondPerformanceTests testUnbindAllPerformance]' measured [Time, seconds] average: 0.000, relative standard deviation: 12.268%, values: [0.000357, 0.000256, 0.000254, 0.000251, 0.000250, 0.000249, 0.000249, 0.000248, 0.000249, 0.000249], performanceMetricID:com.apple.XCTPerformanceMetric_WallClockTime, baselineName: "", baselineAverage: , maxPercentRegression: 10.000%, maxPercentRelativeStandardDeviation: 10.000%, maxRegression: 0.100, maxStandardDeviation: 0.100
Test Case '-[BondTests.DynamicPerformanceTests testDispatchPerformance]' measured [Time, seconds] average: 0.000, relative standard deviation: 52.782%, values: [0.000034, 0.000014, 0.000012, 0.000011, 0.000011, 0.000011, 0.000010, 0.000011, 0.000011, 0.000010], performanceMetricID:com.apple.XCTPerformanceMetric_WallClockTime, baselineName: "", baselineAverage: , maxPercentRegression: 10.000%, maxPercentRelativeStandardDeviation: 10.000%, maxRegression: 0.100, maxStandardDeviation: 0.100
```

After:
```
Test Case '-[BondTests.BondPerformanceTests testBindPerformance]' measured [Time, seconds] average: 0.000, relative standard deviation: 88.803%, values: [0.000036, 0.000011, 0.000010, 0.000007, 0.000008, 0.000006, 0.000006, 0.000005, 0.000007, 0.000005], performanceMetricID:com.apple.XCTPerformanceMetric_WallClockTime, baselineName: "", baselineAverage: , maxPercentRegression: 10.000%, maxPercentRelativeStandardDeviation: 10.000%, maxRegression: 0.100, maxStandardDeviation: 0.100
Test Case '-[BondTests.BondPerformanceTests testUnbindAllPerformance]' measured [Time, seconds] average: 0.000, relative standard deviation: 10.071%, values: [0.000258, 0.000194, 0.000191, 0.000191, 0.000191, 0.000189, 0.000190, 0.000189, 0.000202, 0.000194], performanceMetricID:com.apple.XCTPerformanceMetric_WallClockTime, baselineName: "", baselineAverage: , maxPercentRegression: 10.000%, maxPercentRelativeStandardDeviation: 10.000%, maxRegression: 0.100, maxStandardDeviation: 0.100
Test Case '-[BondTests.DynamicPerformanceTests testDispatchPerformance]' measured [Time, seconds] average: 0.000, relative standard deviation: 57.606%, values: [0.000019, 0.000008, 0.000006, 0.000007, 0.000015, 0.000006, 0.000005, 0.000005, 0.000005, 0.000004], performanceMetricID:com.apple.XCTPerformanceMetric_WallClockTime, baselineName: "", baselineAverage: , maxPercentRegression: 10.000%, maxPercentRelativeStandardDeviation: 10.000%, maxRegression: 0.100, maxStandardDeviation: 0.100
```

### CODE REVIEW

**Comment by ivanmoskalev** [comment]
Needs thorough review + testing with examples.

**Comment by ivanmoskalev** [comment]
Done with that. Please review, and I'll do rework if it is needed.

File: Bond/Bond.swift [hunk 1]
```diff
@@ -66,8 +67,8 @@ public class Bond<T> {
   }
   
   public func bind(dynamic: Dynamic<T>, fire: Bool, strongly: Bool) {
-    dynamic.bonds.append(BondBox(self))
-    
+    dynamic.bonds.insert(BondBox(self))
+
     if strongly {
       self.bondedDynamics.append(dynamic)
```
Comment (srdanrasic) [inline Bond/Bond.swift]: We are keeping this as an array? Well, I guess that's OK. We are using it only for retaining and unbinding purposes.

**Comment by srdanrasic** [comment]
This looks like a very good work. I'll go through it once more, but from what I can see there are no issues.

File: Bond/Bond.swift (hunk 1 above)
Comment (ivanmoskalev) [inline Bond/Bond.swift]: I think a counted set is best fit for all reference counting tasks! 

But since the only way one can manipulate these arrays is `bind(...)` and `unbindAll()`, I think an array is OK! Maybe it could cause trouble if there was an individual `unbind()` operation – for example, if user could do so:
```
b.bind(a) // 1st time
<...>
b.bind(a) // 2nd time
<...>
b.unbind(a) // binding dangling
```
But so far it seems this is out of scope of this issue / PR.

**Comment by ivanmoskalev** [comment]
Thank you! A good architecture is always a pleasure to work with! :)
I thought about one thing though: `Dynamic.bonds` is public. Could my change have broken some end users' code? `append` vs. `insert` etc.

File: Bond/Bond.swift (hunk 1 above)
Comment (srdanrasic) [inline Bond/Bond.swift]: Agreed.

**Comment by srdanrasic** [comment]
I'm thinking of making `Dynamic.bonds` property internal. It should have been marked like that already. I see no reason it should be public.
